### PR TITLE
Simpy Resource compatibility

### PR DESCRIPTION
**Pull request by maxfischer2781**
This pull request provides resources of SimPy for issue #21. This pull request includes:

* [x] fixed incorrect ``Event.callbacks`` type from ``List[E]`` to ``List[(E) -> None]``,
* [x] the ``usim.py.resources`` and ``usim.py.resources.base`` modules,
* [x] the [``usim.py.resources.base`` base classes](https://simpy.readthedocs.io/en/latest/api_reference/simpy.resources.html#module-simpy.resources.base),
* [x] the [``usim.py.resources.resource`` submodule and classes](https://simpy.readthedocs.io/en/latest/api_reference/simpy.resources.html#module-simpy.resources.resource) (CPR #50),
* [x] the [``usim.py.resources.container`` submodule and classes](https://simpy.readthedocs.io/en/latest/api_reference/simpy.resources.html#module-simpy.resources.container) (CPR #51),
* [x] the [``usim.py.resources.store`` submodule and classes](https://simpy.readthedocs.io/en/latest/api_reference/simpy.resources.html#module-simpy.resources.store) (CPR #52).

Child Pull Requests for submodules will follow and SHALL be merged into this PR before acceptance.

### CODE REVIEW

File: usim/py/resources/base.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    def get(self) -> Get[T]:
+        """Create a request to get content out of the resource"""
+        return Get(self)
+
+    def _trigger_put(self, get_event):
+        """
+        Trigger all possible put events
+
+        Called when a ``Put`` event was created or a ``Get`` event triggered.
+        This trigger scheme handles a new request to an unconstested resource
```
Comment (eileen-kuehn) [inline usim/py/resources/base.py]:
```suggestion
        This trigger scheme handles a new request to an uncontested resource
```

File: usim/py/resources/base.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+        """
+        triggered = list(takewhile(self._do_put, self.put_queue))
+        self.put_queue = self.put_queue[len(triggered):]
+
+    def _trigger_get(self, put_event):
+        """
+        Trigger all possible put events
+
+        Called when a ``Get`` event was created or a ``Put`` event triggered.
+        This trigger scheme handles a new request to an unconstested resource
```
Comment (eileen-kuehn) [inline usim/py/resources/base.py]:
```suggestion
        This trigger scheme handles a new request to an uncontested resource
```

**Comment by maxfischer2781** [comment]
No major changes compared to the three child pull requests. Fixed one bug and one inconsistency to SimPy.

Since the changes are mostly on documentation, I've enabled the branch on readthedocs. See the [usim.py overview](https://usim.readthedocs.io/en/feature-simpy-resources/source/api/usim.py.html) and [usim.py.resources subpage](https://usim.readthedocs.io/en/feature-simpy-resources/source/api/usim.py.resources.html#module-usim.py.resources).

File: usim/py/resources/base.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+    def cancel(self):
+        """
+        Cancel the request
+
+        Requests should be cancelled when they have not been triggered but
+        are no longer needed. This may happen when a process is interrupted
+        while waiting for a request.
+
+
```
Comment (eileen-kuehn) [inline usim/py/resources/base.py]:
```suggestion
```

File: usim/py/resources/base.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+        Cancelling is idempotent: if the request has already been triggered or canceled,
```
Comment (eileen-kuehn) [inline usim/py/resources/base.py]:
```suggestion
        Cancelling is idempotent: If the request has already been triggered or canceled,
```

File: usim/py/resources/base.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    """
+    #: The type used to create :py:attr:`~.put_queue`
+    PutQueue: ClassVar[Type[Sequence]] = list
+    #: The type used to create :py:attr:`~.get_queue`
+    GetQueue: ClassVar[Type[Sequence]] = list
+
+    def __init__(self, env: Environment, capacity):
+        self._env = env
+        self._capacity = capacity
+        #: outstanding put events
```
Comment (eileen-kuehn) [inline usim/py/resources/base.py]:
```suggestion
        #: pending put events
```

File: usim/py/resources/base.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+        self.put_queue = self.PutQueue()
+        #: outstanding get events
```
Comment (eileen-kuehn) [inline usim/py/resources/base.py]:
```suggestion
        #: pending get events
```

File: usim_pytest/test_usimpy/test_resources.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+                yield env.timeout(0)
+
+        # 10 claims, 2 concurrent => 5 consecutive claiming pairs
+        claims = [env.process(touch_resource()) for _ in range(10)]\
+            + [env.process(hold_resource(1)) for _ in range(10)]
+        yield claims[-1]
+        assert env.now == 5
+
+    @via_usimpy
+    def test_release_idemptotent(self, env):
```
Comment (eileen-kuehn) [inline usim_pytest/test_usimpy/test_resources.py]:
```suggestion
    def test_release_idempotent(self, env):
```

File: docs/source/api/usim.py.resources.rst [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+===============================================
+
+.. py:module:: usim.py.resources
+    :synopsis: Synchronized Resources
+
+.. toctree::
+   :hidden:
+
+μSim replicates all resource types provided by SimPy.
+Resources synchronize processes by sharing or exchanging objects, data or ownership.
```
Comment (eileen-kuehn) [inline docs/source/api/usim.py.resources.rst]:
```suggestion
Resources synchronize processes by sharing or exchanging objects, data, or ownership.
```

File: docs/source/api/usim.py.resources.rst [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+.. autoclass:: usim.py.resources.resource.PriorityRequest
+    :members:
+
+.. autoclass:: usim.py.resources.resource.Preempted
+    :members:
+
+Containers -- Continuous Resource Exchange
+------------------------------------------
+
+A ``Container`` models the exchange of resources between process:
```
Comment (eileen-kuehn) [inline docs/source/api/usim.py.resources.rst]: Do you refer to several process**es** here?

```suggestion
A ``Container`` models the exchange of resources between processes:
```

**Pull request merged**